include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_checkmark)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label,
+                                 const clap_context_menu_accelerator_t  *accelerator,
+                                 bool                                    is_enabled,
+                                 bool                                    is_checked,
+                                 clap_id                                 action_id);
+
+   bool(CLAP_ABI *add_separator)(const struct clap_context_menu_builder *builder);
+
+   // Each begin_submenu() must have a corresponding end_submenu()
+   bool(CLAP_ABI *begin_submenu)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label);
+
+   bool(CLAP_ABI *end_submenu)(const struct clap_context_menu_builder *builder);
+} clap_context_menu_builder_t;
+
+typedef struct clap_plugin_context_menu {
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *populate)(const clap_plugin_t               *plugin,
+                            const clap_context_menu_target_t  *target,
+                            const clap_context_menu_builder_t *builder);
+
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *perform)(const clap_plugin_t              *plugin,
+                           const clap_context_menu_target_t *target,
+                           clap_id                           action_id);
+} clap_plugin_context_menu_t;
+
+typedef struct clap_host_context_menu {
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *populate)(const clap_host_t                 *host,
+                            const clap_context_menu_target_t  *target,
+                            const clap_context_menu_builder_t *builder);
+
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *perform)(const clap_host_t                *host,
+                           const clap_context_menu_target_t *target,
+                           clap_id action_id);
+
+   // [main-thread]
+   bool(CLAP_ABI *can_popup)(const clap_host_t *host);
+
+   // Shows the host popup menu for a given parameter.
+   // If the plugin is using embedded GUI, then x and y are relative to the plugin's window,
+   // otherwise they're absolute coordinate, and screen index might be set accordingly.
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *popup)(const clap_host_t                *host,
+                         const clap_context_menu_target_t *target,
+                         int32_t                           screen_index,
+                         int32_t                           x,
+                         int32_t                           y);
+} clap_host_context_menu_t;
```

Comment:
Unless I'm missing something, this API is meant for the plugin to request the host to show a context menu *on top* of its plugin GUI?

May I ask, when/why would a plugin trigger that?
Don't plugins that have their own GUI also have their own menus?

In case the plugin is meant to trigger a host menu popup in response to key presses or mouse button presses that the plugin received, the API lacks: a) the mouse button index (1,2,3); b) a time stamp.
Those two a crucial to correctly implement press-drag-release vs click-select popup menu behavior. Also the time stamp is crucial to sort out pointer capture logic (the X11 API has more details on this).

I'm not sure what exactly is needed on mac/win/wayland for this though, and I wonder if it's really worth opening this can of worms (mixing host menus with plugin GUIs), but then I'm assuming plugins with GUIs would have their own menus anyway...

Reply:
I was under the impression that it would be a context menu native to the host's UI, as pictured below. I'm not sure when the "global" menu would pop up, but I'm assuming the "param" menu will show up when you right-click a param knob in Bitwig's UI. I guess it's up to the host what they decide to do with the information, but as I understand it, this isn't intended to be used with a plugin's own UI. To that end, I'm not 100% sure what the purpose of the `popup` method in `clap_host_context_menu_t` is, but I'm guessing it has something to do with embedded UIs.
![context menu](https://user-images.githubusercontent.com/23268982/207850440-c8978551-7815-4846-aaef-de5360baaf09.png)

Reply:
Already asked and answered here: https://github.com/free-audio/clap/pull/221#discussion_r1040583076

So maybe there should be explanatory comment in the code about what's the purpose/use case of the popu9 method?

Reply:
popup as it names suggest is to be called by the plugin to ask the host to popup its menu for a given target.
The host can call plugin->populate() to fetch menu entries.

Reply:
> Don't plugins that have their own GUI also have their own menus?

Most don't. The only VST3 plugin I know of that integrates the host's context menu into its own context menu is Surge XT. But there are many VST3 plugins that pop up the host's context menu when you right click on a parameter, because that's relatively simple to implement and adds a lot of utility to a plugin using a gesture that would otherwise be unused.

Reply:
> popup as it names suggest is to be called by the plugin to ask the host to popup its menu for a given target. The host can call plugin->populate() to fetch menu entries.

Can the host call `plugin->populate()` without a previous call to popup()?

That way the host can offer plugin specific actions in the host device panel menus without the plugin implementing or showing its own GUI.